File: test/parity.ts
/**
 * TS half of the offline test suite. Asserts:
 *  - schema validation rejects the classic bad policies (generic name, bad date,
 *    unknown key, modes/retired overlap) and accepts the real routes.json
 *  - the provider-pin wire mapping matches the shared fixtures
 *  - constraint verdicts match the shared fixtures
 *  - the allowlist and refusal behavior match routes.json
 * No network, no API key. Run with: npx tsx test/parity.ts
 */
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { validateRoutes } from '../src/schema.ts';
import { providerToHttp } from '../src/wire.ts';
import { canDelegate } from '../src/preflight.ts';

const HERE = dirname(fileURLToPath(import.meta.url));
const ROOT = join(HERE, '..');
const fx = JSON.parse(readFileSync(join(HERE, 'fixtures.json'), 'utf8'));
const routes = JSON.parse(readFileSync(join(ROOT, 'routes.json'), 'utf8'));

let failed = 0;
const eq = (a: unknown, b: unknown) => JSON.stringify(a) === JSON.stringify(b);
function ok(name: string, cond: boolean, detail = '') {
  if (cond) {
    console.log(`  ✓ ${name}`);
  } else {
    console.error(`  ✗ ${name}${detail ? ` — ${detail}` : ''}`);
    failed++;
  }
}

console.log('schema validation');
ok('real routes.json validates', validateRoutes(routes).ok);
ok('generic mode name rejected', !validateRoutes({ modes: { chat: mode() } }).ok);
ok(
  'bad verified_date rejected',
  !validateRoutes({ modes: { 'extract-x': mode({ verified_date: 'nope' }) } }).ok,
);
ok('unknown key rejected', !validateRoutes({ modes: { 'extract-x': { ...mode(), typo: 1 } } }).ok);
ok(
  'modes/retired overlap rejected',
  !validateRoutes({ modes: { 'extract-x': mode() }, retired: { 'extract-x': { reason: 'r' } } }).ok,
);
ok(
  'missing required field rejected',
  !validateRoutes({ modes: { 'extract-x': { model: 'm' } } }).ok,
);

console.log('provider-pin wire mapping');
for (const c of fx.provider_pins) {
  ok(
    `pin ${JSON.stringify(c.pin)}`,
    eq(providerToHttp(c.pin), c.http),
    `got ${JSON.stringify(providerToHttp(c.pin))}`,
  );
}

console.log('constraint verdicts');
for (const c of fx.constraint_cases) {
  const got = canDelegate({ constraints: c.constraints } as any, c.meta).ok;
  ok(
    `${JSON.stringify(c.constraints)} vs ${JSON.stringify(c.meta)} → ${c.ok}`,
    got === c.ok,
    `got ${got}`,
  );
}

console.log('allowlist + refusals');
for (const m of fx.refusals.must_exist) ok(`mode "${m}" present`, m in routes.modes);
for (const m of fx.refusals.unknown) ok(`mode "${m}" absent (would refuse)`, !(m in routes.modes));

function mode(over: Record<string, unknown> = {}) {
  return {
    model: 'x',
    priceIn: 0,
    priceOut: 0,
    purpose: 'p',
    use_when: 'u',
    do_not_use_when: 'd',
    verified_date: '2026-01-01',
    ...over,
  };
}

console.log(failed ? `\n✗ TS parity: ${failed} failure(s)` : '\n✓ TS parity: all passed');
process.exit(failed ? 1 : 0);

File: test/parity.py
#!/usr/bin/env python3
"""
Python half of the offline test suite — the mirror of test/parity.ts. Asserts the
Python port agrees with the shared fixtures: same provider-pin wire mapping, same
constraint verdicts, same allowlist + refusals, plus fail-closed behavior and the
embedded-fallback consistency check. No network, no API key.

    python test/parity.py
"""
import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT / "src"))
import router as r  # noqa: E402

fx = json.loads((ROOT / "test" / "fixtures.json").read_text())
routes = json.loads((ROOT / "routes.json").read_text())

failed = 0


def ok(name, cond, detail=""):
    global failed
    if cond:
        print(f"  ✓ {name}")
    else:
        print(f"  ✗ {name}{f' — {detail}' if detail else ''}")
        failed += 1


print("provider-pin wire mapping")
for c in fx["provider_pins"]:
    got = r._provider_http(c["pin"])
    ok(f"pin {json.dumps(c['pin'])}", got == c["http"], f"got {json.dumps(got)}")

print("constraint verdicts")
for c in fx["constraint_cases"]:
    got = r.check_constraints(c["constraints"], c["meta"])["ok"]
    ok(f"{json.dumps(c['constraints'])} vs {json.dumps(c['meta'])} -> {c['ok']}", got == c["ok"], f"got {got}")

print("allowlist + refusals")
for m in fx["refusals"]["must_exist"]:
    ok(f'mode "{m}" resolves', r._model_for(m) == routes["modes"][m]["model"])
for m in fx["refusals"]["unknown"]:
    try:
        r._model_for(m)
        ok(f'mode "{m}" refused', False, "did not raise")
    except ValueError:
        ok(f'mode "{m}" refused', True)

print("embedded-fallback consistency")
cons_ok, msg = r.routes_consistency()
ok("embedded matches canonical routes.json", cons_ok, msg)

print("\n✗ Python parity: %d failure(s)" % failed if failed else "\n✓ Python parity: all passed")
sys.exit(1 if failed else 0)
